packages/tempus/test/TestModels/CDR_Model_impl.hpp

```diff
@@ -449,7 +449,6 @@ void CDR_Model<Scalar>::evalModelImpl(
             }
           }
           if (nonnull(M_inv)) {
-            TEUCHOS_TEST_FOR_EXCEPTION(true,std::runtime_error,"CDR_Model: Preconditioner is NOT implemented for transient term yet!!!");
```

Comment:
If we remove this exception check, then we should fix the preconditioner evaluation that it was checking. It's really simple - just copy the Jacobian evaluation block into this block. If you want me to do this, I can update the pr later today.

Reply:
Uhm, I thought the idea of M_inv was to precondition (M/dt+J) with diag(M)/dt, which kind of makes sense if dt is small. If I replace the computation in there with the one right above for the jacobian, then I would get diag(J) as a preconditioner, which does not really make sense to me.

Reply:
Uhm, now that I look at the code more carefully, I'm confused. I thought the idea behind M_inv was to precondition (M/dt+J) with diag(M)/dt, which makes sense if dt is small. However, I now see that 1) J already contains the mass term, 2) M_inv contains the diffusion/reaction part of the jacobian. I guess the M is one of the letters used to denote preconditioners, while I thought it meant 'mass', whoops.

I can replace the assmebly of M_inv with that of J. Or actually, I can just assembly J, then once it's filled, grab its diagonal. No need to assemble the same matrix twice.

Reply:
agreed. There should probably be a check in the block at the bottom of the evaluate function to make sure that J is nonnull when M_inv is evaluated. The reason for the repeated block was so that you didn't need J to be supplied as part of the evaluate call. I didn't bother to complete this at the time - as I put in the exception to fail out if someone tried to use this capability.

Reply:
Ah, but M_inv may be nonnull even if J is null (though this puzzles me). Ok, need to assemble them separately, I guess.

Reply:
This is just a flexible unit test. Not how I would implement a real physics. More for testing combinations. Where such a case would occur is with Jacobian-Free Newton Krylov. Evaluate the residual and maybe an approximate preconditioner but no Jacobian.

Reply:
That makes sense. Thanks.

Fix commit coming right up.